### ISSUE TITLE
Multithreaded fence tracing is either broken or can be misleading 

### ISSUE THREAD

**Issue by sesse** [apitrace/apitrace#446]
Hi,

I'm having issues with replaying a multithreaded OpenGL trace that heavily involves fences. After a lot of back and forth, I ended up instrumenting Mesa to find the real pointers involved, and I've noticed this happen during replay (after about 1.3M calls):

`
1340124 @4 glFenceSync(condition = GL_SYNC_GPU_COMMANDS_COMPLETE, flags = 0) = 0x7
Mesa warning: inserting 0x7f4e600a3250 in 0x247c0f8
[…stuff happens, nothing of it related to fence 0x7 or 0x4…]
1340429 @1 glDeleteSync(sync = 0x4)
Mesa warning: removing 0x7f4e600a3250 from 0x247c0f8
`

So somehow during the course of the trace, the pointers got mixed up so that both 0x4 and 0x7 refer to the same object. The problem is that there is no kind of warning for this; neither during tracing nor during replay. You only see the errors much later when the 0x7 fence is attempted deleted (and it happens on two entirely different drivers, so I don't think the error is in the driver).

Of course, an even more interesting question is how this came to be in the trace. Among the 1M+ fences that were ever set during the trace, it's not unlikely that the same pointer value was used twice, but that doesn't explain the error. I only ever set and delete fences through a construction with std::shared_ptr, which would make it unlikely (although of course not impossible) to have a double-delete on my side.

Two random theories not really backed by facts:

* Is _sync_map properly protected by a mutex (during tracing and replay)? I couldn't find out exactly how locking works. In particular, if there's a race during tracing between glSyncFence and glDeleteFence and glDeleteFence wins internally (so that glSyncFence can return exactly the pointer that was just deleted, but might return to application code later), what happens?
* Does _sync_map understand that fences are shared between (sharing) contexts?

**Comment by jrfonseca** [comment]
> So somehow during the course of the trace, the pointers got mixed up so that both 0x4 and 0x7 refer to the same object. The problem is that there is no kind of warning for this; neither during tracing nor during replay. 

You can debug the swizzling by running `glretrace -v -v`.  Unfortunately it will dump so much info that might be impracticle for the sort of traces you have.

> Is _sync_map properly protected by a mutex (during tracing and replay)? I couldn't find out exactly how locking works. 

In glretrace, during replay, there's only one active thread at any given instant.  This is by design, so calls happen in the same order as the original trace.  Therefore, no object, global or not, needs any thread protection.

> In particular, if there's a race during tracing between glSyncFence and glDeleteFence and glDeleteFence wins internally (so that glSyncFence can return exactly the pointer that was just deleted, but might return to application code later), what happens?

That race can never happen, as these calls can never happen simultaneously.

> Does _sync_map understand that fences are shared between (sharing) contexts?

There's one _sync_map for all contexts.

Fence objects are shared by all contexts in the same share group.

It's not exactly the same, but as only your apps has only one context share group (and all tend to do, at least I haven't come across one who wants GL context that can't share anything.), it should work.

**Comment by sesse** [comment]
Why can the calls never happen simultaneously? (During trace, that is.)

**Comment by jrfonseca** [comment]
During trace, they can indeed happen simultaneously.   But when replaying, glretrace enforces a order.  See issue #401.

So you're right: it's possible that a race condition that happen during tracing, cause problems when a strict ordering is enforced on the calls during replay.

Entries are never deleted from `_sync_map`, only added.  So if glFenceSync/glDeleteSync happened simultaneously while tracing, and glretrace does it in the wrong order, it might end up deleting the wrong object.

Still, I don't see how that could explain how that could lead to calls you pasted in the opening comment..

I suspect the problem may be more subtle.

**Comment by sesse** [comment]
Well, I do suspect that what I found in my opening comment is itself a follow-up to another root problem. It would be interesting to know what the first instance of operating on a (supposedly) deleted object is in that trace… 

Maybe what I want to search for in my trace is GL_ALREADY_SIGNALED. I should never wait for an object twice, except maybe if something in a lib does it way early. Optionally, I can send you the trace—it's large (3.8 GB) but you only need the first 100 MB or so of it to see these specific issues.

**Comment by jrfonseca** [comment]
> Optionally, I can send you the trace—it's large (3.8 GB) but you only need the first 100 MB or so of it to see these specific issues.

I don't mind taking a look to the trace.  It's probably more effective than speculating.

Size is no problem.  If you don't want to share it publicly see https://github.com/apitrace/apitrace/blob/master/docs/BUGS.markdown#proprietaryconfidential-applications

**Comment by sesse** [comment]
It's a GPL project, so thankfully no issues. You can find the trace at http://storage.sesse.net/nageru.1.trace.gz . Let me know if you have any questions.

Thanks! I'll keep looking on this side as well.

**Comment by sesse** [comment]
I think this is strange; it involves the 0x4 fence that has the wrong (?) address in the original comment:

<pre>
1339885 @4 glMemoryBarrier(barriers = GL_TEXTURE_UPDATE_BARRIER_BIT | GL_CLIENT_MAPPED_BUFFER_BARRIER_BIT)
1339887 @4 glFenceSync(condition = GL_SYNC_GPU_COMMANDS_COMPLETE, flags = 0) = 0x4
Mesa warning: inserting 0x7f47b41061f0 in 0x2af20f8
1339729 @0 glDeleteSync(sync = 0x4)
Mesa warning: removing 0x7f47b41061f0 from 0x2af20f8
1339891 @0 glWaitSync(sync = 0xb, flags = 0, timeout = 18446744073709551615)
</pre>

I didn't delete any lines from the log between these. Why is …729 so out of order? There are deletes at …728 and …734 (both returning before it), if it's relevant.

**Comment by jrfonseca** [comment]
> Why is …729 so out of order? 

Hmm...  ...729 is out of order because, even though `glDeleteSync` call started on event ...729, it actually only finished between ...887 and ...891.  See https://github.com/apitrace/apitrace/issues/401#issuecomment-162341770 for detailed explanation.

> There are deletes at …728 and …734 (both returning before it), if it's relevant.

This is indeed a problem.  Inside OpenGL implementation, the deletion/allocation of sync objects to integer names has a definite order.

But the order apitrace replay calls doesn't match.   We could replay the calls on the order the "enter-function" event happens, but we could have the reverse problem.  We could replay the calls in parallel, but we would just have a race condition.

The only way to fix this is enforce a strict ordering when recording too, ie., holding a global mutex when calling functions that allocate/destroy OpenGL names....

You stumbled across a fundamental flaw in apitrace.  I'll need to think a bit more about this.  

In the meanwhile, you can hack around this yourself in your app by holding mutex around the glFenceSync/DeleteSync calls.

**Comment by sesse** [comment]
> You stumbled across a fundamental flaw in apitrace. I'll need to think a bit more about this.

Achievement unlocked! :-)

TBH what I'd prefer personally is to replay the calls in parallel, but enter in the right order. (The simple reason is that if there's a locking bug in the driver, parallel calls are the only way to have a chance at reproducing it. I realize most people might not want apitrace to find bugs in their drivers, though.) Of course, this doesn't solve the problem of figuring out the trace ordering at all.

In this specific case, I think maybe you can get around it with some custom logic (if the same name is deleted in one thread and created in another at the same time, it's pretty obvious what must have happened), but I don't know if it's possible to get that to scale to the more general problem, especially while keeping overhead reasonably low.

**Comment by sesse** [comment]
So, the best I can think of is this (I'm sorry if it's blindingly obvious):

When you leave a call, you know which other calls are still active (you can presumably keep track of this with some data structure that you update on entry and exit; shouldn't be too heavyweight). You can also keep track of which names they got in (store on entry when you're recording the call anyway). If your call returns a new object that's already “in flight” from another call, you know that you're seeing the wrong order of call retirement, and can either swap the timestamps of the two calls, or (probably easier) spin a bit until the other call retires, since you know that it's already done all its work anyway.

As usual, concurrency is hard, and it's easy to miss some ordering that would topple your scheme… Of course, automatic object creation on bind makes this hard, and it's also an open question if you want to try to guarantee behavior of the user sends in an already-deleted object in one thread while that object happens to be created in another one (I would guess no).

**Comment by jrfonseca** [comment]
What I was thinking was of enforcing a strict ordering for calls to function pairs, by holding a mutex.  Not a single global mutex, but one mutex per namespace:
* a `_sync_mtx` global mutex would be held while calling glFenceSync/glDeleteSync 
* a `_tex_mtx` for glGenTextures/glDeleteTextures
and so on, for all namespaces that are shared across multiple contexts.  Context-specific namespace don't need this.

A slightly better scheme would be to have read/write- mutex, allowing concurrent glFenceSync calls, or concurrent glDeleteSync calls, but not concurrent glFenceSync/glDeleteSync calls.

Keeping track of what's happening on other calls seems too complex.

>TBH what I'd prefer personally is to replay the calls in parallel, but enter in the right order. (The simple reason is that if there's a locking bug in the driver, parallel calls are the only way to have a chance at reproducing it. I realize most people might not want apitrace to find bugs in their drivers, though.) 

Maybe not most of apitrace users, but certainly most of apitrace contributors (myself included) use more apitrace to find bugs in drivers than in apps.  That said, most of the bugs we care about are rendering bugs.

There can certainly be multi-threaded bugs in OpenGL drivers, but my experience is that they are relatively rare.  Partly due to the OpenGL semantics (contexts are not thread safe) partly due to apps that using OpenGL from multiple threads being relatively rare (one rendering thread for every window/drawable) is more common.

Also, using too much parallelism while retracing actually creates more problems for us than solves.  While replaying a trace there is no application doing any synchronization across threads.  Blindly trying to run things in parallel can quite easily cause race-conditions that the real app would never do.  It's simply unpractical.

I understand this might be at odds at your particular goals, as you seem particular interested in the parallelism.  Though, from what you wrote so far, it's quite possible that the random crash you see is not due to multiple CPU threads, but due to CPU <-> GPU parallelism, ie, GPU getting too ahead or too behind and hitting bugs due to improper CPU/GPU synchronization inside OpenGL driver.  In my experience, this is by _far_ more common.  (That said, the way applications are using OpenGL is admitedly changing as result of the widespread availability of multiple cores.)

**Comment by sesse** [comment]
I certainly understand that you can't run parallelism freely across the threads. My intention was only to run the calls in parallel that the trace unambiguously say were run in parallel to begin with.

As for the specific bug I'm trying to hunt down, one can of course only speculate, but the exact same design found a bug in Mesa where fences were not properly internally synchronized. Actually now that the same piece of code has found bugs in Mesa, AMD's drivers, (probably) NVIDIA's driver and now also apitrace, maybe it's time to dial back on the parallelism a bit. :-) (PS: These rare bugs are pretty near impossible to reproduce reliably without something like apitrace. I've been hunting for the “donate” button a few times.)

FWIW, after I changed to a global lock around the glFenceSync and glDeleteSync calls (although not glClientWaitSync), I've run my application for ~12 hours now with no crash. It's a bit too early to say if it's a coincidence or not, though, but it could very well be that the very serialization of these calls masks the bug.

**Comment by sesse** [comment]
OK, it died after 12 hours (and a few minutes) of running, in the exact same spot—free() corruption in glVertexAttribPointer. But while I was looking at the trace, the disk died (!).

So now I know that it _wasn't_ the fence/delete calls running in parallel that caused it, but I have a new possible confounding factor to deal with. Not the least that my trace is on a possibly dead disk. :-P

**Comment by sesse** [comment]
So, new disks in and everything recovered, and I've hit another apitrace issue. My new trace has this:

<pre>
264832628 @3 glBindBuffer(target = GL_PIXEL_PACK_BUFFER, buffer = 0)
264832630 @3 glMemoryBarrier(barriers = GL_TEXTURE_UPDATE_BARRIER_BIT | GL_CLIENT_MAPPED_BUFFER_BARRIER_BIT)
264832632 @3 glFenceSync(condition = GL_SYNC_GPU_COMMANDS_COMPLETE, flags = 0) = 0x7
[nothing more from @3, only some other calls in @0]
264832634 @1 glClientWaitSync(sync = 0x7, flags = 0x0, timeout = 1000000000) = GL_CONDITION_SATISFIED
</pre>

and there it hangs forever. From reading the code, it seems it wants to spin until the sync finishes (since it returned GL_CONDITION_SATISFIED, the remaining trace could depend on that behavior), except that it will never finish, since @3 hasn't flushed yet and won't be flushing since it's not given any more commands.

I suppose I can workaround this in my program by always calling glFlush() after glFenceSync() (before giving the fence to another thread!), but that means I need to grab yet another trace and hope the problem shows up a bit faster this time. Should perhaps apitrace in this case force the other threads to do a flush, to avoid this kind of deadlock?

**Comment by jrfonseca** [comment]
I read https://www.opengl.org/registry/specs/ARB/sync.txt and indeed FenceSync does not do an implicit Flush:

    If the sync object being blocked upon will not be signaled in finite
    time (for example, by an associated fence command issued previously,
    but not yet flushed to the graphics pipeline), then ClientWaitSync
    may hang forever. 

> Should perhaps apitrace in this case force the other threads to do a flush, to avoid this kind of deadlock?

Yes, to be safe, apitrace should indeed invoke glFlush when it switches away from a thread.

It sounds a bit risky thing for an app to do -- to block on one thread without flushing on the thread that emitted the sync --, but as long as the thread that emits flushes eventually (eg, keeps rendering indefinetely or calls glFlush before it stops), then it should indeed be OK. 

a7caf011c7ae5a6d8fd65165db0e1b1c62851659 should fix it.  Give it a go.

It's a bit of an heavy hammer -- it will flush preemptively when switching threads, as opposed to flushing lazily before blocking on WaitSync,  but I don't have yet the plumbing to wake up all threads and do a flush.

**Comment by sesse** [comment]
Yes, the other thread renders indefinitely. In my case, thinking of it, I probably want the frame immediately and not indefinitely, though, so I should add a flush anyway, but it's not necessarily risky; you could easily imagine running with a short timeout and just dealing with the data later if the fence isn't set, but of course apitrace can't model this. I'll test your patch, but right now, my machine is busy trying to take another thread with the flush added :-)

FWIW, you're not guaranteed a flush even in the same thread. There's a special flush flag (with semantics of a special, partial flush) to glClientWaitSync that you can use if the fence was set in the same thread.

**Comment by fred41** [comment]
Considering the way how records are replayed by 'glretrace', we perhaps have a case here, where my suggestion to order the records (in a second pass) by record numbers (by entry time), could be helpful (see #401).

**Comment by sesse** [comment]
@fred41: @jrfonseca has already outlined why it won't help, using that very bug as reference.

**Comment by sesse** [comment]
Strange, when replaying a trace I made with a global mutex around FenceSync/DeleteSync with the new apitrace, I get GL_INVALID_VALUE again. Unsure if they were there with the old one too, or if they just drowned in all the noise. But either locking isn't enough to get the timing in apitrace right, or I've done something wrong in my implementation of it.

**Comment by sesse** [comment]
OK, I really don't get it. Here's my trace:

<pre>
10046384 @2 glBindTexture(target = GL_TEXTURE_2D, texture = 0)
10046386 @2 glBindBuffer(target = GL_PIXEL_UNPACK_BUFFER, buffer = 0)
10046388 @2 glFenceSync(condition = GL_SYNC_GPU_COMMANDS_COMPLETE, flags = 0) = 0xa
10046391 @3 glWaitSync(sync = 0x5, flags = 0, timeout = 18446744073709551615)
10046393 @3 glDeleteSync(sync = 0x5)
10046396 @3 glWaitSync(sync = 0xa, flags = 0, timeout = 18446744073709551615)
10046398 @3 glDeleteSync(sync = 0xa)
10046403 @3 glDisable(cap = GL_DITHER)
10046405 @3 glEnable(cap = GL_FRAMEBUFFER_SRGB)
</pre>

And the error message I see is

<pre>
10046396: message: major api error 1281: GL_INVALID_VALUE error generated. <sync> is not the name of a sync object.
10046396 @3 glWaitSync(sync = 0xa, flags = 0, timeout = 18446744073709551615)
10046396: warning: glGetError(glWaitSync) = GL_INVALID_VALUE
10046398: message: major api error 1281: GL_INVALID_VALUE error generated. <sync> is neither zero nor the name of a sync object.
10046398 @3 glDeleteSync(sync = 0xa)
10046398: warning: glGetError(glDeleteSync) = GL_INVALID_VALUE
</pre>

Surely 0xa should be valid, coming from …388? Unless something went wrong with the interpretation of 0x5.

4.7 GB of glorious trace (compressed to 2.5 GB using gzip) at http://storage.sesse.net/nageru.17.trace.gz .

**Comment by jrfonseca** [comment]
Note that glretrace keeps swizzling sync names behind your back. The messages you see show the sync that was recorded, not the sync that was used when replayed.  To see the later run `glretrace -v -v ` and look for messages with ` <- ` and ` -> `, as these show which values are these being mapped to.  And it's possible there's yet some bug somewhere.

**Comment by sesse** [comment]
At least I replayed the trace to the end, and it… doesn't crash. Like the previous trace, it ends in several OpenGL commands being run at the same time:

<pre>
615823428 @4 glTexSubImage2D(target = GL_TEXTURE_2D, level = 0, xoffset = 0, yoffset = 0, width = 1280, height = 720, format = GL_RED, type = GL_UNSIGNED_BYTE, pixels = 0x400000) // incomplete
615823471 @0 glVertexAttribPointer(index = 1, size = 2, type = GL_FLOAT, normalized = GL_FALSE, stride = 0, pointer = NULL) // incomplete
615823473 @2 glBindBuffer(target = GL_PIXEL_UNPACK_BUFFER, buffer = 156) // incomplete
</pre>

Especially the fact that glVertexAttribPointer() is the dying call here too, and that there's a glBindBuffer() at the same time, really makes me think it's a concurrency bug. I've got more apitrace runs going to see if I can collect more data.
